### PR TITLE
Migration from Karma to Jest

### PR DESCRIPTION
**Pull request by pstjean**
As part of preparing for #347, I knew I wanted to tighten the [feedback loop](https://github.com/testdouble/contributing-tests/wiki/Feedback-Loop) significantly on the test suite. I knew Jest was massively faster than any other JS test runner out there (especially with the benefit of parallelism on this dual core hyperthreaded processor I've got). I started playing with this idea in transit to and from Boston, running the codemods and poking around. I finished it off yesterday with tweaks to support JSDOM's peculiarities, a wonky import or two and a fix to styleguide.

This PR reduces the run time of the test suite by over 70%, from **58 seconds** to **16 seconds**. Additionally, Jest has a wonderful [smart watch mode](https://frontendmasters.com/courses/testing-react/jest-watch-mode/) that keeps an eye on your Git status and only runs tests relevant to files that have changed, which typically means only a small percentage of your test suite needs to be run each time you make a code change. Hopefully, this translates to:
1. 80%+ less waiting around for the test suite to run
2. Much less time between writing and validating code

Since Jest uses JSDOM instead of a headless browser for DOM-dependent tests, this also removes Phantom from our dependencies, which should speed up builds.

### CODE REVIEW

**Comment by cavis** [comment]
Woooooo no more phantomjs!

File: src/app/shared/menu/date/custom-date-range-dropdown.component.spec.ts [hunk 1]
```diff
@@ -118,9 +118,12 @@ describe('CustomDateRangeDropdownComponent', () => {
     expect(comp.googleAnalyticsEvent).toHaveBeenCalledWith(comp.CUSTOM_DATE, comp.tempRange);
   });
 
-  it('should not close the dropdown on window scroll', () => {
+  it('should not close the dropdown on window scroll', done => {
     comp.open = true;
+    window.addEventListener('scroll', (e) => {
```
Comment (sandikbarr) [inline src/app/shared/menu/date/custom-date-range-dropdown.component.spec.ts]: how does addEventListener also dispatch the event, is that a jest thing...?

File: src/app/ngrx/effects/routing.effects.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { Injectable } from '@angular/core';
 import { map } from 'rxjs/operators/map';
 import { switchMap } from 'rxjs/operators/switchMap';
-import { Observable } from 'rxjs/Observable';
+import { Observable } from 'rxjs';
```
[the remark this answers is not in the document]
Comment (sandikbarr) [reply src/app/ngrx/effects/routing.effects.ts]: Running the tests, looks like the import problem was with `Observable.of`. If we import of `import { of } from 'rxjs/observable/of';` then change `Observable.of` to just `of` that should fix it and save us the bundle size increase.

File: src/app/shared/menu/date/custom-date-range-dropdown.component.spec.ts (hunk 1 above)
Comment (sandikbarr) [inline src/app/shared/menu/date/custom-date-range-dropdown.component.spec.ts]: it honestly makes a bit more sense that we'd need some kind of asynchronous test on these actually

File: package.json [hunk 3]
```diff
@@ -5,11 +5,11 @@
   "scripts": {
```
[the remark this answers is not in the document]
Comment (sandikbarr) [reply package.json]: weird, sorry I missed that yarn.lock was in here, not sure why I had changes in mine locally, but I wont worry about it

File: src/app/ngrx/effects/routing.effects.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (pstjean) [reply src/app/ngrx/effects/routing.effects.ts]: Yikes, good to know!

As far as I can tell, `rxjs/Observable` is just a type definition file. So I think we need to import both `of` in the way that you suggested, and `rxjs/Observable` for the type. Pushed that fix.

I don't really understand how the Angular build works. I'm guessing some special inference from the type definition to discover and import the code you need when you reference a static member of that type in the way we were doing there. 🤷‍♂️

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (pstjean) [reply package.json]: I made sure that was included when I bumped styleguide just now.

**Pull request merged**